File: scripts/fetch_calendar.py
import json
import re
import sys
from datetime import datetime, timedelta
from datetime import date as date_from_datetime

import requests
import scrape_id

# Get school name
SCHOOL_NAME = input("Enter school name: ")
output_file = f"../public/calendars/{SCHOOL_NAME}.json"

try:
    SITE_ID = scrape_id.get_site_info(SCHOOL_NAME)["site_id"]
except (RuntimeError, KeyError):
    print(f"{SCHOOL_NAME} not found.")
    sys.exit(1)

# Configuration
SCHOOL_YEAR = "2026-2027"

START_DATE = "2026-09-01"
END_DATE = "2027-06-30"

# Hijack lrsd school calendar URLs
DISTRICT_URL = "https://cicmsapi.azurewebsites.net/lrsd/_ci/15/ci/vsb/webservice.ashx"
EVENT_URL = DISTRICT_URL + "/" + SCHOOL_NAME

HEADERS = {
    "Accept": "*/*",
    "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8",
    "Origin": "https://www.lrsd.net",
    "Referer": "https://www.lrsd.net/",
}

# Generate every day
calendar = {}

current = datetime.strptime(START_DATE, "%Y-%m-%d")
end = datetime.strptime(END_DATE, "%Y-%m-%d")

while current <= end:
    date = current.strftime("%Y-%m-%d")
    weekday = current.weekday()  # Monday = 0, Sunday = 6

    has_school = weekday < 5

    calendar[date] = {
        "date": date,
        "hasSchool": has_school,
        "timeSlot": "Regular",
        "status": "Normal School Day" if has_school else "No School",
        "holidays": [],
        "dayInfo": []
    }

    current += timedelta(days=1)

# Download school events
search = {
    "Keyword": "",
    "Category": "",
    "StartDate": START_DATE,
    "EndDate": END_DATE,
    "SelectedChildren": [],
    "SelectedCalendars": [],
    "SchoolClasses": [],
    "ForMonthView": True,
    "NoRecurExpand": False,
    "SiteId": SITE_ID,
    "CategoryGuids": []
}

event_payload = {
    "CategoryName": "EventsAdvancedSerach",
    "MethodName": "Search",
    "Parameters": {
        "json": json.dumps(search, separators=(",", ":"))
    }
}

print("Fetching calendar data...")
events_response = requests.post(
    EVENT_URL,
    headers=HEADERS,
    data={
        "ansp": json.dumps(event_payload, separators=(",", ":"))
    }
)

events_response.raise_for_status()

events = events_response.json()["Events"]
print("Calendar data aquired.")

# Keyword Matching via Regex
NO_SCHOOL_PATTERNS = [
    r"\bno school\b",
    r"\bno classes\b",
    r"\b(winter|spring|christmas|fall) break\b",
    r"\bchristmas day\b",
    r"\bboxing day\b",
    r"\bnew year'?s? day\b",
    r"\bvictoria day\b",
    r"\bthanksgiving\b",
    r"\bfamily day\b",
    r"\blabour day\b",
    r"\bgood friday\b",
    r"\bremembrance day\b",
    r"\bcanada day\b",
    r"\bheritage day\b",
    r"\bholiday\b",
    r"\b(in[- ]?service|pd day|professional development)\b", # Catches PD days
]

# Allowlist for events that mention holidays but are actually school days
SCHOOL_BUT_KEYWORD_MATCH = [
    "last day of classes before",
    "first day back",
    "return from",
    "classes resume"
]

def is_no_school(title):
    t = title.lower()
    if any(phrase in t for phrase in SCHOOL_BUT_KEYWORD_MATCH):
        return False
    return any(re.search(p, t) for p in NO_SCHOOL_PATTERNS)

EARLY_DISMISSAL_KEYWORDS = ["early dismissal"]

# Iterate through multi-day events
for event in events:
    start_str = event.get("StartTime", "")[:10]
    end_str = event.get("EndTime", "")[:10]
    
    if not start_str:
        continue

    start_date = date_from_datetime.fromisoformat(start_str)
    end_date = date_from_datetime.fromisoformat(end_str) if end_str else start_date
    
    # If EndTime is somehow before StartTime, just use StartTime
    if end_date < start_date:
        end_date = start_date

    cur_date = start_date
    while cur_date <= end_date:
        date_str = cur_date.isoformat()
        
        if date_str in calendar:
            entry = calendar[date_str]
            title = event.get("Title", "").strip()
            lower = title.lower()

            if title and title not in entry["dayInfo"]:
                entry["dayInfo"].append(title)

            if is_no_school(title):
                entry["hasSchool"] = False
                entry["status"] = "No School"
                entry["timeSlot"] = "Regular"

                if title not in entry["holidays"]:
                    entry["holidays"].append(title)

            elif any(word in lower for word in EARLY_DISMISSAL_KEYWORDS):
                # Only apply early dismissal if the day isn't already marked "No School"
                if entry["hasSchool"]:
                    entry["status"] = "Early Dismissal"
                    entry["timeSlot"] = "Early Dismissal"

        cur_date += timedelta(days=1)

# Save
calendar = dict(sorted(calendar.items()))

for date_str, info in calendar.items():
    if not info["hasSchool"]:
        continue
    if info["holidays"]:
        continue
        
    dt = date_from_datetime.fromisoformat(date_str)
    if dt.weekday() == 2:  # Wednesday
        info["dayInfo"].append("Late Start Remembrance Day")

with open(output_file, "w", encoding="utf-8") as f:
    json.dump(calendar, f, indent=2)

print(f"Wrote {len(calendar)} days to {output_file}")

File: src/fetchCalendar.ts
import rawSchoolData from "./schools.json";

interface School {
  codeName: string;
  id: number;
  preBuilt: boolean;
  highSchool: boolean;
}

interface SchoolObj {
  [schoolName: string]: School;
}

export const schoolData = rawSchoolData as unknown as SchoolObj;

interface CalendarDay {
  date: string;
  hasSchool: boolean;
  timeSlot: "Regular" | "Early Dismissal";
  status: "Normal School Day" | "No School" | "Early Dismissal";
  holidays: Array<string>;
  dayInfo: Array<string>;
}

interface CalendarIndex {
  [date: string]: CalendarDay;
}

interface EventItem {
  Title?: string;
  StartTime?: string;
  EndTime?: string;
  [key: string]: unknown;
}

interface EventsAPIResponse {
  Events: Array<EventItem>;
}

// Configuration
const START_DATE = "2026-09-05";
const END_DATE = "2027-06-30";

// Hijack lrsd school calendar URLs
const DISTRICT_URL =
  "https://cicmsapi.azurewebsites.net/lrsd/_ci/15/ci/vsb/webservice.ashx";

const HEADERS = {
  "Accept": "*/*",
  "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8",
  "Origin": "https://www.lrsd.net",
  "Referer": "https://www.lrsd.net/",
};

function toISODate(d: Date): string {
  const year = d.getUTCFullYear();
  const month = String(d.getUTCMonth() + 1).padStart(2, "0");
  const day = String(d.getUTCDate()).padStart(2, "0");
  return `${year}-${month}-${day}`;
}

function getSchoolId(schoolName: string): number {
  const schoolObj = Object.values(schoolData).find((school) => school.codeName === schoolName)
  if (!schoolObj) {
    throw new RangeError(`${schoolName} does not exist.`);
  }
  return schoolObj.id;
}

export async function getCalendar(schoolName: string): Promise<CalendarIndex> {
  const eventURL = `${DISTRICT_URL}/${schoolName}`;

  const siteID = getSchoolId(schoolName);

  const calendar: CalendarIndex = {};

  const current = new Date(`${START_DATE}T00:00:00Z`);
  const end = new Date(`${END_DATE}T00:00:00Z`);

  while (current.getTime() <= end.getTime()) {
    const date = toISODate(current);
    const weekday = current.getUTCDay();

    const hasSchool = weekday >= 1 && weekday <= 5;

    calendar[date] = {
      date,
      hasSchool,
      timeSlot: "Regular",
      status: hasSchool ? "Normal School Day" : "No School",
      holidays: [],
      dayInfo: [],
    };

    current.setUTCDate(current.getUTCDate() + 1);
  }

  const search = {
    "Keyword": "",
    "Category": "",
    "StartDate": START_DATE,
    "EndDate": END_DATE,
    "SelectedChildren": [],
    "SelectedCalendars": [],
    "SchoolClasses": [],
    "ForMonthView": true,
    "NoRecurExpand": false,
    "SiteId": siteID,
    "CategoryGuids": [],
  };

  const eventPayload = {
    "CategoryName": "EventsAdvancedSerach",
    "MethodName": "Search",
    "Parameters": {
      "json": JSON.stringify(search),
    },
  };

  const eventsResponse = await fetch(eventURL, {
    method: "POST",
    headers: HEADERS,
    body: new URLSearchParams({
      ansp: JSON.stringify(eventPayload),
    }),
  });

  if (!eventsResponse.ok) {
    throw new Error(`HTTP error! status: ${eventsResponse.status}`);
  }

  const data = (await eventsResponse.json()) as EventsAPIResponse;
  const events = data.Events;

  // Keyword matching via regex
  const noSchoolPatterns: RegExp[] = [
    /\bno school\b/,
    /\bno classes\b/,
    /\b(winter|spring|christmas|fall) break\b/,
    /\bchristmas day\b/,
    /\bboxing day\b/,
    /\bnew year'?s? day\b/,
    /\bvictoria day\b/,
    /\bthanksgiving\b/,
    /\bfamily day\b/,
    /\blabour day\b/,
    /\bgood friday\b/,
    /\bremembrance day\b/,
    /\bcanada day\b/,
    /\bheritage day\b/,
    /\bholiday\b/,
    /\b(in[- ]?service|pd day|professional development)\b/,
  ];

  // Allowlist for events that mention holidays but are actually school days
  const schoolButKeywordMatch = [
    "last day of classes before",
    "first day back",
    "return from",
    "classes resume",
  ];

  function isNoSchool(title: string): boolean {
    const t = title.toLowerCase();
    if (schoolButKeywordMatch.some((phrase) => t.includes(phrase))) {
      return false;
    }
    return noSchoolPatterns.some((p) => p.test(t));
  }

  const earlyDismissalKeywords = ["early dismissal"];

  // Iterate through multi-day events
  for (const event of events) {
    const startStr = (event.StartTime ?? "").slice(0, 10);
    const endStr = (event.EndTime ?? "").slice(0, 10);

    if (!startStr) {
      continue;
    }

    // Date-only ISO strings parse as UTC midnight
    const startDate = new Date(startStr);
    let endDate = endStr ? new Date(endStr) : new Date(startDate.getTime());

    // If EndTime is somehow before StartTime, just use StartTime
    if (endDate.getTime() < startDate.getTime()) {
      endDate = new Date(startDate.getTime());
    }

    const curDate = new Date(startDate.getTime());
    while (curDate.getTime() <= endDate.getTime()) {
      const dateStr = toISODate(curDate);

      const entry = calendar[dateStr];
      if (entry) {
        const title = (event.Title ?? "").trim();
        const lower = title.toLowerCase();

        if (title && !entry.dayInfo.includes(title)) {
          entry.dayInfo.push(title);
        }

        if (isNoSchool(title)) {
          entry.hasSchool = false;
          entry.status = "No School";
          entry.timeSlot = "Regular";

          if (!entry.holidays.includes(title)) {
            entry.holidays.push(title);
          }
        } else if (earlyDismissalKeywords.some((word) => lower.includes(word))) {
          // Only apply early dismissal if the day isn't already marked "No School"
          if (entry.hasSchool) {
            entry.status = "Early Dismissal";
            entry.timeSlot = "Early Dismissal";
          }
        }
      }

      curDate.setUTCDate(curDate.getUTCDate() + 1);
    }
  }

  // Save (sorted)
  const sorted = Object.fromEntries(
    Object.entries(calendar).sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0))
  );

  for (const info of Object.values(sorted)) {
    if (!info.hasSchool) {
      continue;
    }
    if (info.holidays.length > 0) {
      continue;
    }

    const dt = new Date(info.date);
    if (dt.getUTCDay() === 3) {
      info.dayInfo.push("Late Start Remembrance Day");
    }
  }

  return sorted;
}
